//-*-mode:protobuf; mode:font-lock;-*-

/*

   calin/proto/math/histogram.proto -- Stephen Fegan -- 2015-06-11

   Protobufs for representing histograms

   Copyright 2015, Stephen Fegan <sfegan@llr.in2p3.fr>
   LLR, Ecole Polytechnique, CNRS/IN2P3

   This file is part of "calin"

   "calin" is free software: you can redistribute it and/or modify it
   under the terms of the GNU General Public License version 2 or
   later, as published by the Free Software Foundation.

   "calin" is distributed in the hope that it will be useful, but
   WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   General Public License for more details.

*/

syntax = "proto3";
option cc_enable_arenas = true;

import "calin.proto";

package calin.ix.math.histogram;

message Histogram1DConfig {
  double dxval                          = 1
    [(CFO).desc = "Bin width", (CFO).units = "xval-units" ];
  double xval_align                     = 2
    [(CFO).desc = "Bin alignment, such that (x0/dx mod 1)=(xalign mod 1), where x0 is "
      "the x-value at the left hand side of the first (or any) bin.",
     (CFO).units = "bins" ];
  bool limited                          = 5
    [(CFO).desc = "Limit the histogram range" ];
  double xval_limit_lo                  = 6
    [(CFO).desc = "Lower limit on X value", (CFO).units = "xval-units" ];
  double xval_limit_hi                  = 7
    [(CFO).desc = "Upper limit on X value", (CFO).units = "xval-units" ];
  string name                           = 13
    [(CFO).desc = "Name of the histogram"];
  string xval_units                     = 14
    [(CFO).desc = "X-value units"];
  string weight_units                   = 15
    [(CFO).desc = "Weight units"];
};

message Histogram1DData {
  option (CMO).message_integration_function = MIF_USER_GENERATED;

  double dxval                          = 1
    [(CFO).desc = "Bin width", (CFO).units = "xval-units" ];
  double xval_align                     = 2
    [(CFO).desc = "Bin alignment, such that (x0 mod dx)=xalign",
     (CFO).units = "bins" ];
  double xval0                          = 3
    [(CFO).desc = "X-value of LHS of first bin", (CFO).units = "xval-units" ];
  repeated double bins                  = 4
    [(CFO).desc = "Weights for the bins", (CFO).units = "weight-units" ];
  bool limited                          = 5
    [(CFO).desc = "Limit the histogram range" ];
  double xval_limit_lo                  = 6
    [(CFO).desc = "Lower limit on X value", (CFO).units = "xval-units" ];
  double xval_limit_hi                  = 7
    [(CFO).desc = "Upper limit on X value", (CFO).units = "xval-units" ];
  double overflow_lo                    = 8
    [(CFO).desc = "Total weight that overflows lower limit",
     (CFO).units = "weight-units" ];
  double overflow_hi                    = 9
    [(CFO).desc = "Total weight that overflows upper limit",
     (CFO).units = "weight-units" ];
  double sum_w                          = 10
    [(CFO).desc = "Total weight in histogram",
     (CFO).units = "weight-units" ];
  double sum_wx                         = 11
    [(CFO).desc = "Total x-val times weight in histogram",
     (CFO).units = "weight-units xval-units" ];
  double sum_wxx                        = 12
    [(CFO).desc = "Total x-val squared times weight in histogram",
     (CFO).units = "weight-units xval-units^2" ];
  string name                           = 13
    [(CFO).desc = "Name of the histogram"];
  string xval_units                     = 14
    [(CFO).desc = "X-value units"];
  string weight_units                   = 15
    [(CFO).desc = "Weight units"];
};
